### PR TITLE
[OptApp] Add NLOPT Library to OptimizationApplication

### PR DESCRIPTION
**Pull request by RezaNajian**
This PR adds support for the NLOPT optimization library in the OptimizationApplication of Kratos Multiphysics. NLOPT is a popular open-source library for nonlinear optimization, providing a wide range of optimization algorithms. By integrating NLOPT into the OptimizationApplication, Kratos users can now access its powerful optimization capabilities.

Please mark the PR with appropriate tags:

Enhancement, OptimizationApplication, ExternalLibraries

🆕 Changelog

Added NLOPT as an external library in custom_external_libraries
Implemented NLOPT support in the OptimizationApplication
Added examples and tests for using NLOPT in the OptimizationApplication

### CODE REVIEW

File: applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    std::copy(optimizer->m_initial_guess.begin(), optimizer->m_initial_guess.end(), grad);
+}
+
+void NLOptOptimizer::InitializeOptimizer()
+{
+    m_opt = nlopt_create(m_algorithm, m_num_variables);
+
+    if (m_opt == nullptr) {
+        throw std::runtime_error("Could not create NLOpt optimizer");
+    }
```
Comment (philbucher) [inline applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.cpp]:
```suggestion
    KRATOS_ERROR_IF_NOT(m_opt) << "Could not create NLOpt optimizer" << std::endl;
```
please use the provided error macros

File: applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    if (grad != nullptr && optimizer->m_gradient != nullptr) {
+        optimizer->m_gradient(x_vec, optimizer->m_initial_guess);
+        std::copy(optimizer->m_initial_guess.begin(), optimizer->m_initial_guess.end(), grad);
+    }
+
+    return result;
+}
+
+void NLOptOptimizer::GradientWrapper(unsigned int n, const double* x, double* grad, void* data)
```
Comment (philbucher) [inline applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.cpp]: I would avoid working with `void *` at all cost

Why do you use pointers?

File: applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    nlopt_set_lower_bounds(m_opt, m_lower_bounds.data());
+    nlopt_set_upper_bounds(m_opt, m_upper_bounds.data());
+    nlopt_set_initial_step(m_opt, m_initial_guess.data());
+    nlopt_set_xtol_rel(m_opt, m_relative_tolerance);
+    nlopt_set_ftol_abs(m_opt, m_absolute_tolerance);
+    nlopt_set_maxeval(m_opt, m_max_iterations);
+
+    if (m_gradient != nullptr) {
+        //nlopt_set_gradient(m_opt, GradientWrapper, this);
+    }
```
Comment (philbucher) [inline applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.cpp]:
```suggestion
    if (m_gradient) {
        //nlopt_set_gradient(m_opt, GradientWrapper, this);
    }
```
this is modern C++

File: applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            static double ObjectiveFunctionWrapper(unsigned n, const double* x, double* grad, void* data);
+            static void GradientWrapper(unsigned n, const double* x, double* grad, void* data);
+            void InitializeOptimizer();
+
+            nlopt_algorithm m_algorithm;
+            unsigned int m_num_variables;
+            nlopt_opt m_opt;
+            std::function<double(std::vector<double>&)> m_objective_function;
+            std::function<void(std::vector<double>&, std::vector<double>&)> m_gradient;
+            std::vector<double> m_lower_bounds;
```
Comment (philbucher) [inline applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.h]:
```suggestion
            std::vector<double> mLowerBounds;
```

please follow the style guide

File: applications/OptimizationApplication/custom_external_libraries/nlopt/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+list (APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
+
+option (NLOPT_CXX "enable cxx routines" ON)
+option (NLOPT_FORTRAN "enable fortran tests" OFF)
+option (BUILD_SHARED_LIBS "Build NLopt as a shared library" ON)
+option (NLOPT_PYTHON "build python bindings" ON)
+option (NLOPT_OCTAVE "build octave bindings" ON)
+option (NLOPT_MATLAB "build matlab bindings" ON)
+option (NLOPT_GUILE "build guile bindings" ON)
+option (NLOPT_SWIG "use SWIG to build bindings" ON)
```
Comment (philbucher) [inline applications/OptimizationApplication/custom_external_libraries/nlopt/CMakeLists.txt:45]: are those needed?

Can they be set to `OFF`?

File: applications/OptimizationApplication/custom_external_libraries/nlopt/COPYING [hunk 6]
```diff
@@ -0,0 +1,40 @@
+NLopt combines several free/open-source nonlinear optimization
+libraries by various authors.  See the COPYING, COPYRIGHT, and README
+files in the subdirectories for the original copyright and licensing
+information of these packages.
+
+The compiled NLopt library, i.e. the combined work of all of the
+included optimization routines, is licensed under the conjunction of
+all of these licensing terms.  Currently, the most restrictive terms
+are for the code in the "luksan" directory, which is licensed under
+the GNU Lesser General Public License (GNU LGPL), version 2.1 or
```
Comment (philbucher) [inline applications/OptimizationApplication/custom_external_libraries/nlopt/COPYING:10]: This is not compatible with the license of Kratos

**Review by philbucher — COMMENTED** [review]
some minor comments

The license might not be compatible with Kratos, can you please comment?

In any case, adding external libs needs approval from @KratosMultiphysics/technical-committee 

Alternatively you could add the lib as optional dependency, i.e. with a CMake flag, then you dont have to include the source (e.g. like we do for Trilinos, Metis, MKL ...)

File: applications/OptimizationApplication/custom_external_libraries/nlopt/COPYING (hunk 6 above)
Comment (loumalouomega) [reply applications/OptimizationApplication/custom_external_libraries/nlopt/COPYING:10]: It is the same as MMG, please follow same approach

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
To be discussed by the @KratosMultiphysics/technical-committee.

As @loumalouomega suggests, I'd follow the same approach we did for the MMG. In this regard, I stress that the library shouldn't be included in our repo but installed by the user.

**Comment by loumalouomega** [comment]
> I'm also agreeing with @rubenzorrilla. We can use the `FindCmake` file to find the library and link it. so the user has the responsibility to install and manage it. There won't be copyright issues.
> 
> In that case we can make these lib inclusions optional.

If `FindCmake` I would update other dependencies as well for  sake of consistency and simplicity.

**Comment by roigcarlo** [comment]
As others said, it would be better if you can add it as an optional dependency. Just to add that, as for the compatibility goes, MIT and BSD4 are fine.

File: applications/OptimizationApplication/custom_external_libraries/nlopt/COPYING (hunk 6 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/custom_external_libraries/nlopt/COPYING:10]: yes. Thanks for the hint. I will check and follow MMG work.

File: applications/OptimizationApplication/custom_external_libraries/nlopt/CMakeLists.txt (hunk 5 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/custom_external_libraries/nlopt/CMakeLists.txt:45]: yes actually :-)

File: applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.h (hunk 4 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/custom_algorithms/nlopt_optimizer.h]: sure, however this is the draft to show case the idea and work.

**Pull request closed without merging**